packages/loader/container-loader/src/container.ts

```diff
@@ -987,21 +976,61 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         return this._deltaManager.connect(args);
     }
 
+    public async finishLoad(request: IRequest) {
+        if (this._context !== undefined) { return; }
+
+        return PerformanceEvent.timedExecAsync(this.logger, { eventName: "Load" }, async (event) => {
+            // This code to capture errors from "closed" event is only needed
+            // - for safety, in case error does not propagate through stack
+            // - for safety, to break out any infinite retries by closing container.
+            // - for cases where loader.request() is used directly by caller, not leveraging loader.resolve()
+            //   to get container earlier and be able to register all the right handlers to listen for errors.
+            return new Promise<Container>((res, rej) => {
+                const onClosed = (err?: ICriticalContainerError) => {
+                    rej(err ?? CreateContainerError("Container closed without an error"));
+                };
+                this.on("closed", onClosed);
+
+                this._attachState = AttachState.Attached;
```

Comment:
Why are we setting attached state twice? Here and in create? We usually set attached state after connecting to storage service.